### PR TITLE
impl(common): decorator to log REST credentials

### PR DESCRIPTION
**Pull request by coryan**
This will be used in the initialization of
`oauth2_internal::Credentials`.  I am planning to log any usage of cached credentials, and also any refreshes.  Using `ComputeEngineCredentials` as an example, and with all types implicitly in the `oauth2_internal` namespace, the full typical stack would be:

```
LoggingCredentials(cached) -> CachedCredentials ->
     LoggingCredentials(refresh) -> ComputeEngineCredentials
```

The logging elements are optional. They will be inserted only if the `g::c::TracingComponentsOption` includes `auth`.

Part of the work for #10316 and for #3429

- - -
This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/10408)

### CODE REVIEW

File: google/cloud/internal/oauth2_logging_credentials.cc [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+                   << token.status();
+    return token;
+  }
+  if (now <= token->expiration) {
+    GCP_LOG(DEBUG) << __func__ << "(" << phase_ << "), token=" << *token
+                   << ", token will expire in "
+                   << absl::FormatDuration(
+                          absl::FromChrono(token->expiration - now));
+    return token;
+  }
```
Comment (devbww) [inline google/cloud/internal/oauth2_logging_credentials.cc:48]: Nit: I'd make this "good" case the final one by inverting the test.

File: google/cloud/internal/oauth2_logging_credentials.h [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+namespace oauth2_internal {
+GOOGLE_CLOUD_CPP_INLINE_NAMESPACE_BEGIN
+
+/**
+ * Log credentials usage and refreshes.
+ *
+ * Debugging problems with authentication can be fairly difficult. This
+ * decorator is automatically inserted (twice) if
+ * `google::cloud::TracingComponentsOption` includes `auth`. The decorator is
+ * inserted before and after the caching layer, to show whether an cached token
```
Comment (devbww) [inline google/cloud/internal/oauth2_logging_credentials.h:33]: s/an cached/a cached/

File: google/cloud/internal/oauth2_logging_credentials.h [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+ * @note To prevent leaking authentication secrets, the tokens are not logged in
+ * full.
+ *
+ * @see https://cloud.google.com/docs/authentication/ for an overview of
+ * authenticating to Google Cloud Platform APIs.
+ */
+class LoggingCredentials : public Credentials {
+ public:
+  LoggingCredentials(std::string phase, TracingOptions tracing_options,
+                     std::shared_ptr<Credentials> impl);
```
Comment (devbww) [inline google/cloud/internal/oauth2_logging_credentials.h:45]: `#include <memory>`

File: google/cloud/internal/oauth2_logging_credentials.h [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+ *
+ * @note To prevent leaking authentication secrets, the tokens are not logged in
+ * full.
+ *
+ * @see https://cloud.google.com/docs/authentication/ for an overview of
+ * authenticating to Google Cloud Platform APIs.
+ */
+class LoggingCredentials : public Credentials {
+ public:
+  LoggingCredentials(std::string phase, TracingOptions tracing_options,
```
Comment (devbww) [inline google/cloud/internal/oauth2_logging_credentials.h:44]: `#include <string>`

File: google/cloud/internal/oauth2_logging_credentials.h [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+  StatusOr<internal::AccessToken> GetToken(
+      std::chrono::system_clock::time_point now) override;
+  StatusOr<std::vector<std::uint8_t>> SignBlob(
+      absl::optional<std::string> const& signing_service_account,
+      std::string const& string_to_sign) const override;
+  std::string AccountEmail() const override;
+  std::string KeyId() const override;
+
+ private:
+  std::string const phase_;
```
Comment (devbww) [inline google/cloud/internal/oauth2_logging_credentials.h:57]: Nit: We usually think (I think) the `const` isn't worth sacrificing copyability (or, if you actually don't want copyability, to do that explicitly).

Comment (coryan) [reply google/cloud/internal/oauth2_logging_credentials.h]: Fixed

File: google/cloud/internal/oauth2_logging_credentials.h (hunk 4 above)
Comment (coryan) [reply google/cloud/internal/oauth2_logging_credentials.h:46]: Fixed

File: google/cloud/internal/oauth2_logging_credentials.cc (hunk 1 above)
Comment (coryan) [reply google/cloud/internal/oauth2_logging_credentials.cc]: Done

File: google/cloud/internal/oauth2_logging_credentials.h (hunk 2 above)
Comment (coryan) [reply google/cloud/internal/oauth2_logging_credentials.h]: Fixed, thanks

File: google/cloud/internal/oauth2_logging_credentials.h (hunk 3 above)
Comment (coryan) [reply google/cloud/internal/oauth2_logging_credentials.h:47]: Fixed

**Review by devbww — APPROVED** [review]

**Pull request merged**